In [1]:
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.multioutput import ClassifierChain
from sklearn.svm import LinearSVC
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
import pandas as pd




from sklearn.datasets import fetch_openml
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, accuracy_score




In [2]:
mnist = fetch_openml('mnist_784', as_frame = False, parser = 'auto') #return Numpy arrays. to get a dataframe set as_frame = True
X, y = mnist.data, mnist.target

In [3]:
Xtrain, Xdev, Xtest = X[:50000], X[50000:60000], X[60000:]
ytrain, ydev, ytest = y[:50000], y[50000:60000], y[60000:]


In [4]:
#KNeighborsClassifier

#KNN classifier 
knn_classifier = KNeighborsClassifier(n_neighbors=5)

#Flatten the images
Xtrain_flat = Xtrain.reshape((Xtrain.shape[0], -1))
Xdev_flat = Xdev.reshape((Xdev.shape[0], -1))
Xtest_flat = Xtest.reshape((Xtest.shape[0], -1))

# Fit the classifier on the flattened training data
knn_classifier.fit(Xtrain_flat, ytrain)

#predictions on the development (validation) set
ydev_pred = knn_classifier.predict(Xdev_flat)

#accuracy on the development set
accuracy_knn = accuracy_score(ydev, ydev_pred)
print(f"Accuracy on the dev set: {accuracy_knn:.2%}")

#predictions on the test set
ytest_pred = knn_classifier.predict(Xtest_flat)

#accuracy on the test set
accuracy_test = accuracy_score(ytest, ytest_pred)
print(f"Accuracy on the test set: {accuracy_test:.2%}")

#F1 score on the development set
f1_knn = f1_score(ydev, ydev_pred, average='weighted')
print(f"F1 score on the dev set: {f1_knn:.2%}")
#F1 score on the test set
f1_test = f1_score(ytest, ytest_pred, average='weighted')
print(f"F1 score on the test set: {f1_test:.2%}")

#recall score on the development set
recall_knn = recall_score(ydev, ydev_pred, average='weighted')
print(f"Recall on the dev set: {recall_knn:.2%}")
#recall score on the test set
recall_test = recall_score(ytest, ytest_pred, average='weighted')
print(f"Recall on the test set: {recall_test:.2%}")

#precision score on the development set
precision_knn = precision_score(ydev, ydev_pred, average='weighted')
print(f"Precision on the dev set: {precision_knn:.2%}")

#precision score on the test set
precision_test = precision_score(ytest, ytest_pred, average='weighted')
print(f"Precision on the test set: {precision_test:.2%}")




Accuracy on the dev set: 97.18%
Accuracy on the test set: 96.64%
F1 score on the dev set: 97.17%
F1 score on the test set: 96.63%
Recall on the dev set: 97.18%
Recall on the test set: 96.64%
Precision on the dev set: 97.20%
Precision on the test set: 96.67%


In [5]:
#SGD Classifier

sgd_classifier = SGDClassifier(random_state=42)

Xtrain_flat = Xtrain.reshape((Xtrain.shape[0], -1))
Xdev_flat = Xdev.reshape((Xdev.shape[0], -1))
Xtest_flat = Xtest.reshape((Xtest.shape[0], -1))

sgd_classifier.fit(Xtrain_flat, ytrain)

ydev_pred = sgd_classifier.predict(Xdev_flat)

accuracy_sgd = accuracy_score(ydev, ydev_pred)
print(f"Accuracy on the dev set: {accuracy_sgd:.2%}")

ytest_pred = sgd_classifier.predict(Xtest_flat)

accuracy_test = accuracy_score(ytest, ytest_pred)
print(f"Accuracy on the test set: {accuracy_test:.2%}")
#F1
f1_sgd = f1_score(ydev, ydev_pred, average='weighted')
print(f"F1 score on the SGD dev set: {f1_sgd:.2%}")
f1_test = f1_score(ytest, ytest_pred, average='weighted')
print(f"F1 score on the test set: {f1_test:.2%}")

#Recall
recall_sgd = recall_score(ydev, ydev_pred, average='weighted')
print(f"Recall on the SGD dev set: {recall_sgd:.2%}")
recall_test = recall_score(ytest, ytest_pred, average='weighted')
print(f"Recall on the SGD test set: {recall_test:.2%}")

#Precision

precision_sgd = precision_score(ydev, ydev_pred, average='weighted')
print(f"Precision on the SGD dev set: {precision_sgd:.2%}")
precision_test = precision_score(ytest, ytest_pred, average='weighted')
print(f"Precision on the SGD test set: {precision_test:.2%}")



Accuracy on the dev set: 87.39%
Accuracy on the test set: 86.77%
F1 score on the SGD dev set: 87.51%
F1 score on the test set: 86.97%
Recall on the SGD dev set: 87.39%
Recall on the SGD test set: 86.77%
Precision on the SGD dev set: 88.25%
Precision on the SGD test set: 87.99%


In [6]:
#LinearSVC classifier
linear_svc_classifier = LinearSVC(random_state=42)

Xtrain_flat = Xtrain.reshape((Xtrain.shape[0], -1))
Xdev_flat = Xdev.reshape((Xdev.shape[0], -1))
Xtest_flat = Xtest.reshape((Xtest.shape[0], -1))

linear_svc_classifier.fit(Xtrain_flat, ytrain)

ydev_pred = linear_svc_classifier.predict(Xdev_flat)

accuracy_svc = accuracy_score(ydev, ydev_pred)
print(f"Accuracy on the dev set: {accuracy_svc:.2%}")

ytest_pred = linear_svc_classifier.predict(Xtest_flat)

accuracy_test = accuracy_score(ytest, ytest_pred)
print(f"Accuracy on the test set: {accuracy_test:.2%}")



f1_svc = f1_score(ydev, ydev_pred, average='weighted')
print(f"F1 score on the LinearSVC dev set: {f1_svc:.2%}")

f1_test = f1_score(ytest, ytest_pred, average='weighted')
print(f"F1 score on the LinearSVC test set: {f1_test:.2%}")


recall_svc = recall_score(ydev, ydev_pred, average='weighted')
print(f"Recall on the LinearSVC dev set: {recall_svc:.2%}")
recall_test = recall_score(ytest, ytest_pred, average='weighted')
print(f"Recall on the LinearSVC test set: {recall_test:.2%}")


precision_svc = precision_score(ydev, ydev_pred, average='weighted')
print(f"Precision on the LinearSVC dev set: {precision_svc:.2%}")
precision_test = precision_score(ytest, ytest_pred, average='weighted')
print(f"Precision on the LinearSVC test set: {precision_test:.2%}")



/home/LC/manzhe01/.local/lib/python3.8/site-packages/sklearn/svm/_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
/home/LC/manzhe01/.local/lib/python3.8/site-packages/sklearn/svm/_base.py:1250: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy on the dev set: 85.78%
Accuracy on the test set: 85.98%
F1 score on the LinearSVC dev set: 85.83%
F1 score on the LinearSVC test set: 86.00%
Recall on the LinearSVC dev set: 85.78%
Recall on the LinearSVC test set: 85.98%
Precision on the LinearSVC dev set: 86.89%
Precision on the LinearSVC test set: 86.94%


In [7]:
#DecisionTree
decision_tree_classifier = DecisionTreeClassifier(random_state=42)
Xtrain_flat = Xtrain.reshape((Xtrain.shape[0], -1))
Xdev_flat = Xdev.reshape((Xdev.shape[0], -1))
Xtest_flat = Xtest.reshape((Xtest.shape[0], -1))
decision_tree_classifier.fit(Xtrain_flat, ytrain)
ydev_pred = decision_tree_classifier.predict(Xdev_flat)
accuracy_tree = accuracy_score(ydev, ydev_pred)
print(f"Accuracy on the dev set: {accuracy_tree:.2%}")
ytest_pred = decision_tree_classifier.predict(Xtest_flat)
accuracy_test = accuracy_score(ytest, ytest_pred)
print(f"Accuracy on the test set: {accuracy_test:.2%}")
# Make predictions on the development (validation) set
ydev_pred = decision_tree_classifier.predict(Xdev_flat)

#F1 score on the development set
f1_tree = f1_score(ydev, ydev_pred, average='weighted')
print(f"F1 score on the DecisionTree dev set: {f1_tree:.2%}")
#F1 score on the test set
f1_test = f1_score(ytest, ytest_pred, average='weighted')
print(f"F1 score on the DecisionTree test set: {f1_test:.2%}")

#recall score on the development set
recall_tree = recall_score(ydev, ydev_pred, average='weighted')
print(f"Recall on the DecisionTree dev set: {recall_tree:.2%}")
#recall score on the test set
recall_test = recall_score(ytest, ytest_pred, average='weighted')
print(f"Recall on the DecisionTree test set: {recall_test:.2%}")

#precision score on the development set
precision_tree = precision_score(ydev, ydev_pred, average='weighted')
print(f"Precision on the DecisionTree dev set: {precision_tree:.2%}")
#precision score on the test set
precision_test = precision_score(ytest, ytest_pred, average='weighted')
print(f"Precision on the DecisionTree test set: {precision_test:.2%}")

#predictions on the test set
ytest_pred = decision_tree_classifier.predict(Xtest_flat)






Accuracy on the dev set: 88.22%
Accuracy on the test set: 86.93%
F1 score on the DecisionTree dev set: 88.21%
F1 score on the DecisionTree test set: 86.90%
Recall on the DecisionTree dev set: 88.22%
Recall on the DecisionTree test set: 86.93%
Precision on the DecisionTree dev set: 88.21%
Precision on the DecisionTree test set: 86.91%


In [8]:
#Logistic Regression
logistic_regression_classifier = LogisticRegression(random_state=42, max_iter=100)
Xtrain_flat = Xtrain.reshape((Xtrain.shape[0], -1))
Xdev_flat = Xdev.reshape((Xdev.shape[0], -1))
Xtest_flat = Xtest.reshape((Xtest.shape[0], -1))
logistic_regression_classifier.fit(Xtrain_flat, ytrain)
#predictions on the development (validation) set
ydev_pred = logistic_regression_classifier.predict(Xdev_flat)
#predictions on the test set
ytest_pred = logistic_regression_classifier.predict(Xtest_flat)
#accuracy on the test set
accuracy_test = accuracy_score(ytest, ytest_pred)
print(f"Accuracy on the LogisticRegression test set: {accuracy_test:.2%}")
#accuracy on the development set
accuracy_logreg = accuracy_score(ydev, ydev_pred)
print(f"Accuracy on the LogisticRegression dev set: {accuracy_logreg:.2%}")


#predictions on the development (validation) set
ydev_pred = logistic_regression_classifier.predict(Xdev_flat)

#F1 score on the development set
f1_logreg = f1_score(ydev, ydev_pred, average='weighted')
print(f"F1 score on the LogisticRegression dev set: {f1_logreg:.2%}")
#predictions on the test set
ytest_pred = logistic_regression_classifier.predict(Xtest_flat)

#F1 score on the test set
f1_test = f1_score(ytest, ytest_pred, average='weighted')
print(f"F1 score on the LogisticRegression test set: {f1_test:.2%}")

#recall score on the development set
recall_logreg = recall_score(ydev, ydev_pred, average='weighted')
print(f"Recall on the LogisticRegression dev set: {recall_logreg:.2%}")
#recall score on the test set
recall_test = recall_score(ytest, ytest_pred, average='weighted')
print(f"Recall on the LogisticRegression test set: {recall_test:.2%}")

#precision score on the development set
precision_logreg = precision_score(ydev, ydev_pred, average='weighted')
print(f"Precision on the LogisticRegression dev set: {precision_logreg:.2%}")
#precision score on the test set
precision_test = precision_score(ytest, ytest_pred, average='weighted')
print(f"Precision on the LogisticRegression test set: {precision_test:.2%}")


/home/LC/manzhe01/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy on the LogisticRegression test set: 92.43%
Accuracy on the LogisticRegression dev set: 92.72%
F1 score on the LogisticRegression dev set: 92.70%
F1 score on the LogisticRegression test set: 92.41%
Recall on the LogisticRegression dev set: 92.72%
Recall on the LogisticRegression test set: 92.43%
Precision on the LogisticRegression dev set: 92.69%
Precision on the LogisticRegression test set: 92.42%


In [9]:

# Define the scores for each algorithm
scores = {
    'Algorithm': ['SGDClassifier', 'LinearSVC', 'KNeighborsClassifier', 'LogisticRegression', 'DecisionTree'],
    'Accuracy': [accuracy_sgd, accuracy_svc, accuracy_knn, accuracy_logreg, accuracy_tree],
    'Precision': [precision_sgd, precision_svc, precision_knn, precision_logreg, precision_tree],
    'Recall': [recall_sgd, recall_svc, recall_knn, recall_logreg, recall_tree],
    'F1 Score': [f1_sgd, f1_svc, f1_knn, f1_logreg, f1_tree]
}

# Create a Pandas DataFrame from the scores dictionary
df = pd.DataFrame(scores)

# Print the table for comparison
print(df)


              Algorithm  Accuracy  Precision  Recall  F1 Score
0         SGDClassifier    0.8739   0.882472  0.8739  0.875077
1             LinearSVC    0.8578   0.868867  0.8578  0.858331
2  KNeighborsClassifier    0.9718   0.972000  0.9718  0.971737
3    LogisticRegression    0.9272   0.926883  0.9272  0.926951
4          DecisionTree    0.8822   0.882139  0.8822  0.882090


In [16]:
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=42)
X_dev, X_test, y_dev, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

knn = KNeighborsClassifier(n_neighbors=5)  

knn.fit(X_train, y_train)

y_dev_pred = knn.predict(X_dev)

y_test_pred = knn.predict(X_test)

dev_accuracy = accuracy_score(y_dev, y_dev_pred)
print(f"Accuracy on the development set: {dev_accuracy * 100:.2f}%")

test_accuracy = accuracy_score(y_test, y_test_pred)
print(f"Accuracy on the test set: {test_accuracy * 100:.2f}%")


Accuracy on the development set: 96.95%
Accuracy on the test set: 96.73%


We can see that the KNeighborsClassifier  algorithm worked the best because of its accuracy on every score is higher comapred to the other algorithms.

Accuracy: Among the algorithms, KNeighborsClassifier achieved the highest accuracy (97.18%), indicating that it correctly classified a significant portion of the data.

Precision: KNeighborsClassifier also has the highest precision (97.20%), suggesting that when it predicts a positive class, it is accurate.

Recall: KNeighborsClassifier's recall (97.18%) is very high, indicating that it effectively captures most of the positive class instances.

F1 Score: KNeighborsClassifier has the highest F1 score (97.17%), which is a balance between precision and recall. It suggests a strong overall performance.